In [1]:
# Google Drive Mount

from google.colab import auth
auth.authenticate_user()

from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/gdrive


In [2]:
!ls

adc.json  gdrive  sample_data


In [3]:
# Change Directory to Google Drive
cd /content/gdrive/My \Drive/

/content/gdrive/My Drive


In [4]:
!ls

'!대면 인터뷰 피드백(190809) .gdoc'   머신러닝가즈앙.ipynb   VRO
'Colab Notebooks'		      Untitled0.ipynb	     YBIGTA
 CROWDWORKS			      Untitled1.ipynb


In [5]:
cd YBIGTA/ieee-fraud-detection

/content/gdrive/My Drive/YBIGTA/ieee-fraud-detection


In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [0]:
import pandas_profiling
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split

In [0]:
trans = pd.read_csv('train_transaction.csv', index_col='TransactionID')
test_trans = pd.read_csv('test_transaction.csv', index_col='TransactionID')

In [0]:
ident = pd.read_csv('train_identity.csv', index_col='TransactionID')
test_ident = pd.read_csv('test_identity.csv', index_col='TransactionID')

In [0]:
sample_sub = pd.read_csv('sample_submission.csv', index_col='TransactionID')

In [78]:
trans.head()

isFraud  TransactionDT  TransactionAmt  ... V337  V338  V339
TransactionID                                          ...                 
2987000              0          86400            68.5  ...  NaN   NaN   NaN
2987001              0          86401            29.0  ...  NaN   NaN   NaN
2987002              0          86469            59.0  ...  NaN   NaN   NaN
2987003              0          86499            50.0  ...  NaN   NaN   NaN
2987004              0          86506            50.0  ...  0.0   0.0   0.0

[5 rows x 393 columns]

In [12]:
trans.tail(3)

isFraud  TransactionDT  TransactionAmt  ... V337  V338  V339
TransactionID                                          ...                 
3577537              0       15811079           30.95  ...  NaN   NaN   NaN
3577538              0       15811088          117.00  ...  NaN   NaN   NaN
3577539              0       15811131          279.95  ...  NaN   NaN   NaN

[3 rows x 393 columns]

In [13]:
test_trans.head()

TransactionDT  TransactionAmt ProductCD  ...  V337  V338  V339
TransactionID                                           ...                  
3663549             18403224           31.95         W  ...   NaN   NaN   NaN
3663550             18403263           49.00         W  ...   NaN   NaN   NaN
3663551             18403310          171.00         W  ...   NaN   NaN   NaN
3663552             18403310          284.95         W  ...   NaN   NaN   NaN
3663553             18403317           67.95         W  ...   NaN   NaN   NaN

[5 rows x 392 columns]

In [14]:
ident.head()

id_01     id_02  ...  DeviceType                     DeviceInfo
TransactionID                   ...                                           
2987004          0.0   70787.0  ...      mobile  SAMSUNG SM-G892A Build/NRD90M
2987008         -5.0   98945.0  ...      mobile                     iOS Device
2987010         -5.0  191631.0  ...     desktop                        Windows
2987011         -5.0  221832.0  ...     desktop                            NaN
2987016          0.0    7460.0  ...     desktop                          MacOS

[5 rows x 40 columns]

In [15]:
ident.tail(3)

id_01     id_02  ...  DeviceType                          DeviceInfo
TransactionID                   ...                                                
3577529        -20.0  632381.0  ...      mobile  Moto E (4) Plus Build/NMA26.42-152
3577531         -5.0   55528.0  ...     desktop                               MacOS
3577534        -45.0  339406.0  ...      mobile         RNE-L03 Build/HUAWEIRNE-L03

[3 rows x 40 columns]

In [16]:
test_ident.head()

id_01     id_02  ...  DeviceType                   DeviceInfo
TransactionID                   ...                                         
3663586        -45.0  280290.0  ...      mobile  MYA-L13 Build/HUAWEIMYA-L13
3663588          0.0    3579.0  ...      mobile         LGLS676 Build/MXB48T
3663597         -5.0  185210.0  ...     desktop                  Trident/7.0
3663601        -45.0  252944.0  ...      mobile  MYA-L13 Build/HUAWEIMYA-L13
3663602        -95.0  328680.0  ...      mobile         SM-G9650 Build/R16NW

[5 rows x 40 columns]

In [79]:
trans.shape

(590540, 393)

In [18]:
test_trans.shape

(506691, 392)

In [19]:
ident.shape

(144233, 40)

In [20]:
test_ident.shape

(141907, 40)

In [0]:
train = trans.merge(ident, how='left', left_index=True, right_index=True)
test = test_trans.merge(test_ident, how='left', left_index=True, right_index=True)

In [22]:
train.head(3)

isFraud  TransactionDT  ...  DeviceType DeviceInfo
TransactionID                          ...                       
2987000              0          86400  ...         NaN        NaN
2987001              0          86401  ...         NaN        NaN
2987002              0          86469  ...         NaN        NaN

[3 rows x 433 columns]

In [81]:
train.shape

(590540, 433)

In [82]:
test.head(3)

TransactionDT  TransactionAmt  ... DeviceType  DeviceInfo
TransactionID                                 ...                       
3663549             18403224           31.95  ...        NaN         NaN
3663550             18403263           49.00  ...        NaN         NaN
3663551             18403310          171.00  ...        NaN         NaN

[3 rows x 432 columns]

In [83]:
test.shape #isFraud만 빼고 다 있음.

(506691, 432)

In [0]:
pd.options.display.max_rows = 500

In [27]:
(train.isnull().sum()/len(train)*100).sort_values(ascending=False) 
#433개 중 19개 빼고 다 결측치가 있군 

id_24             99.196159
id_25             99.130965
id_07             99.127070
id_08             99.127070
id_21             99.126393
id_26             99.125715
id_22             99.124699
id_23             99.124699
id_27             99.124699
dist2             93.628374
D7                93.409930
id_18             92.360721
D13               89.509263
D14               89.469469
D12               89.041047
id_04             88.768923
id_03             88.768923
D6                87.606767
id_33             87.589494
id_10             87.312290
D8                87.312290
D9                87.312290
id_09             87.312290
id_30             86.865411
id_32             86.861855
id_34             86.824771
id_14             86.445626
V149              86.123717
V153              86.123717
V156              86.123717
V154              86.123717
V155              86.123717
V157              86.123717
V158              86.123717
V161              86.123717
V162              86

In [28]:
print(train.isnull().any().sum()) # train에 결측치 있는 놈들 414개  (433개중)

414


In [0]:
pd.options.display.max_rows = 50

In [30]:
one_value_cols =[col for col in train.columns if train[col].nunique() <= 1]
one_value_cols #0개

[]

In [31]:
test_one_value_cols =[col for col in test.columns if test[col].nunique() <= 1]
test_one_value_cols #1개 

['V107']

In [0]:
bool_cols = [col for col in train if
            train[col].dropna().value_counts().index.isin([0,1]).all()]

In [33]:
bool_cols #binary인 거 이정도밖에 없군

['isFraud', 'V1', 'V14', 'V41', 'V65', 'V88', 'V107']

In [34]:
Fr_or_not = train['isFraud']
Fr_or_not = pd.to_numeric(Fr_or_not, errors='coerce')
corr1 = train[['isFraud', 'V1', 'V14', 'V41', 'V65', 'V88', 'V107']].corrwith(Fr_or_not)
corr1.sort_values(ascending=False)

#혹시나 해봤는데 아직 뭐 처리해주기도 전이니 상관계수 쓸모없음 ~~ isFraud는 당연히 1

isFraud    1.000000
V41        0.004154
V65        0.002701
V107       0.002554
V14        0.002145
V1         0.001045
V88        0.000554
dtype: float64

In [0]:
TF_cols = [col for col in train if
            train[col].dropna().value_counts().index.isin(['T','F']).all()]

In [36]:
TF_cols #M1~3, 5~9는 + id_35~38은 그냥 T:1, F:0식으로 바꿔줘도 편할듯.

['M1',
 'M2',
 'M3',
 'M5',
 'M6',
 'M7',
 'M8',
 'M9',
 'id_35',
 'id_36',
 'id_37',
 'id_38']

In [37]:
print(train['M4'].unique())  #얘만 이렇게 되어 있네,,,

['M2' 'M0' nan 'M1']


In [38]:
print(1-144233/590540)
(train[TF_cols].isnull().sum()/len(train[TF_cols])*100).sort_values(ascending=False) 

0.7557608290716971


id_38    76.126088
id_37    76.126088
id_36    76.126088
id_35    76.126088
M5       59.349409
M7       58.635317
M9       58.633115
M8       58.633115
M3       45.907136
M2       45.907136
M1       45.907136
M6       28.678836
dtype: float64

"""
id cols는 애초에 identity의 행이 transaction의 1/4 크기니까 
최소 75.5%의 결측치가 있음. 따라서 76.13%의 결측치는 보기만 그렇지
id의 거의 대부분의 행이 값을 가지고 있다는 것... 중요한 변수일 느낌? 
"""

In [0]:
FNF_cols = [col for col in train if
            train[col].dropna().value_counts().index.isin(['NotFound','Found']).all()]

In [40]:
FNF_cols

['id_12', 'id_16', 'id_27', 'id_29']

In [41]:
print(1-144233/590540)
(train[FNF_cols].isnull().sum()/len(train[FNF_cols])*100).sort_values(ascending=False) 

0.7557608290716971


id_27    99.124699
id_16    78.098012
id_29    76.127273
id_12    75.576083
dtype: float64

In [42]:
print(train['id_23'].unique()) #Transparent, Anonymous, Hidden --> 값 이름만 보면 중요할 느낌...그저 느낌
print(train['id_28'].unique()) #이자식은 New, Found
print(train['id_34'].unique()) #match_status? 

[nan 'IP_PROXY:TRANSPARENT' 'IP_PROXY:ANONYMOUS' 'IP_PROXY:HIDDEN']
[nan 'New' 'Found']
[nan 'match_status:2' 'match_status:1' 'match_status:0' 'match_status:-1']


In [43]:
train.describe() ##Fraud일 확률 3.49%...! #train.loc[(train['isFraud']==1)]

#1 ) TransactionAmt == 거래되는 돈, 평균 135달러
# --> 최댓값 31937.391000이네 그럼 4천마넌 !! 총 59만 달러,,, 7억 정도밖에 안 되네 머야
#2 ) card의 개별 값들 보면 소수점 없는 걸로 봐서 각각 뭔가의 지표 ex) 카드종류?

isFraud  TransactionDT  ...        id_26         id_32
count  590540.000000   5.905400e+05  ...  5163.000000  77586.000000
mean        0.034990   7.372311e+06  ...   149.070308     26.508597
std         0.183755   4.617224e+06  ...    32.101995      3.737502
min         0.000000   8.640000e+04  ...   100.000000      0.000000
25%         0.000000   3.027058e+06  ...   119.000000     24.000000
50%         0.000000   7.306528e+06  ...   149.000000     24.000000
75%         0.000000   1.124662e+07  ...   169.000000     32.000000
max         1.000000   1.581113e+07  ...   216.000000     32.000000

[8 rows x 402 columns]

In [44]:
train.describe(include=['O'])

ProductCD   card4   card6  ...   id_38 DeviceType DeviceInfo
count     590540  588963  588969  ...  140985     140810     118666
unique         5       4       4  ...       2          2       1786
top            W    visa   debit  ...       F    desktop    Windows
freq      439670  384767  439938  ...   73922      85165      47722

[4 rows x 31 columns]

In [45]:
print(train['M1'].unique())

['T' nan 'F']


In [46]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 590540 entries, 2987000 to 3577539
Columns: 433 entries, isFraud to DeviceInfo
dtypes: float64(399), int64(3), object(31)
memory usage: 1.9+ GB


In [47]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 506691 entries, 3663549 to 4170239
Columns: 432 entries, TransactionDT to DeviceInfo
dtypes: float64(399), int64(2), object(31)
memory usage: 1.7+ GB


In [0]:
sns.countplot(train['ProductCD'])
plt.show() #이거 왜 그래프 안 나오지...

In [49]:
train[['ProductCD','isFraud']].groupby('ProductCD',
as_index=False).mean().sort_values(by='isFraud',ascending=False)
# W가 많긴 많은데 C가 젤 확률이 높넹

ProductCD   isFraud
0         C  0.116873
3         S  0.058996
1         H  0.047662
2         R  0.037826
4         W  0.020399

In [0]:
del trans, test_trans, ident, test_ident

In [51]:
trans.shape

NameError: ignored

In [52]:
print(train.shape) ; test.shape

(590540, 433)


(506691, 432)

In [53]:
print(train.isnull().any().sum())

414


# **내가 처리할 부분 ! V columms**

In [84]:
V_cols = [col for col in train.columns if col.startswith('V')]
train[V_cols].describe() #339개나 있군링

V1             V2  ...           V338           V339
count  311253.000000  311253.000000  ...   82351.000000   82351.000000
mean        0.999945       1.045204  ...     151.160542     100.700882
std         0.007390       0.240133  ...    1095.034387     814.946722
min         0.000000       0.000000  ...       0.000000       0.000000
25%         1.000000       1.000000  ...       0.000000       0.000000
50%         1.000000       1.000000  ...       0.000000       0.000000
75%         1.000000       1.000000  ...       0.000000       0.000000
max         1.000000       8.000000  ...  104060.000000  104060.000000

[8 rows x 339 columns]

In [60]:
print(train[col].nunique() <= 5 for col in V_cols)

<generator object <genexpr> at 0x7f82f81656d0>


In [85]:
del_cols = [col for col in train.columns if train[col].isnull().sum() / train.shape[0] > 0.75]
del_cols_test = [col for col in test.columns if test[col].isnull().sum() / test.shape[0] > 0.75]
print(len(del_cols))
print(len(del_cols_test)) ##결측치 75% 이상인 거 날리자 ~~

208
76


In [86]:
total_del_cols = list(set(del_cols + del_cols_test))
print('is_Fraud' in total_del_cols)
len(total_del_cols)

False


208

In [87]:
len(V_cols)

339

In [0]:
train = train.drop(total_del_cols, axis=1)
test = test.drop(total_del_cols, axis=1)

In [89]:
print(train.shape) ; test.shape

(590540, 225)


(506691, 224)

In [110]:
good_V_cols = [col for col in train.columns if col.startswith('V')]
len(good_V_cols)

180

In [119]:
not_V_cols_ = [col for col in train.columns if not col.startswith('V')]
print('isFraud' in not_V_cols_)
not_V_cols_.pop(0)
not_V_cols_
not_V_cols = not_V_cols_
len(not_V_cols)

True


44

In [0]:
train = train.drop(not_V_cols, axis=1)
test = test.drop(not_V_cols, axis=1)

In [121]:
print(train.shape) ; test.shape

(590540, 181)


(506691, 180)

In [122]:
train.loc[:,train.dtypes =='object'].columns

Index([], dtype='object')

In [123]:
train.loc[:,train.dtypes =='int'].columns

Index(['isFraud'], dtype='object')

In [124]:
train.loc[:,train.dtypes =='float'].columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       ...
       'V312', 'V313', 'V314', 'V315', 'V316', 'V317', 'V318', 'V319', 'V320',
       'V321'],
      dtype='object', length=180)

이제 isFraud 말고는 다 float,,,~~

In [125]:
train.describe()

isFraud             V1  ...           V320           V321
count  590540.000000  311253.000000  ...  590528.000000  590528.000000
mean        0.034990       0.999945  ...      42.073133      28.326584
std         0.183755       0.007390  ...     473.499307     382.053171
min         0.000000       0.000000  ...       0.000000       0.000000
25%         0.000000       1.000000  ...       0.000000       0.000000
50%         0.000000       1.000000  ...       0.000000       0.000000
75%         0.000000       1.000000  ...       0.000000       0.000000
max         1.000000       1.000000  ...  104060.000000  104060.000000

[8 rows x 181 columns]

In [0]:
one_train = train
two_train = train
three_train = train
four_train = train
five_train = train

In [0]:

for col in one_train.columns:
  one_train[col] = one_train[col].fillna(method='ffill') #bfill

In [133]:
final_col = train.columns
len(final_col)

181

In [134]:
one_train[final_col].columns[pd.isnull(one_train[final_col]).sum() != 0]

Index(['V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41', 'V42', 'V43', 'V44',
       'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51', 'V52'],
      dtype='object')

In [136]:
one_train[final_col].columns[pd.isnull(one_train[final_col]).sum() == 0]

Index(['isFraud', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9',
       ...
       'V312', 'V313', 'V314', 'V315', 'V316', 'V317', 'V318', 'V319', 'V320',
       'V321'],
      dtype='object', length=163)

In [137]:
train[final_col].columns[pd.isnull(train[final_col]).sum() == 0]

Index(['isFraud', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9',
       ...
       'V312', 'V313', 'V314', 'V315', 'V316', 'V317', 'V318', 'V319', 'V320',
       'V321'],
      dtype='object', length=163)

In [148]:
has_null_cols = ['V35', 'V36', 'V37', 'V38', 'V39', 'V40', 'V41', 'V42', 'V43', 'V44',
       'V45', 'V46', 'V47', 'V48', 'V49', 'V50', 'V51', 'V52']
len(has_null_cols)

18

In [0]:
for col in has_null_cols:
  one_train[col] = one_train[col].fillna(method='bfill')

In [152]:
one_train[final_col].columns[pd.isnull(one_train[final_col]).sum() != 0]

Index([], dtype='object')

In [153]:
one_train[final_col].columns[pd.isnull(one_train[final_col]).sum() == 0]

Index(['isFraud', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9',
       ...
       'V312', 'V313', 'V314', 'V315', 'V316', 'V317', 'V318', 'V319', 'V320',
       'V321'],
      dtype='object', length=181)

In [158]:
one_train.shape[1]

181

In [0]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC, LinearSVC

In [0]:
kneighbors = KNeighborsClassifier()
kneighbors.fit(one_train[final_col],one_train['isFraud'])
kneighbors_score = round(kneighbors.score(one_train[final_col],one_train['isFraud'])*100, 2)

kneighbors_score

In [0]:
model1 = KNeighborsClassifier(n_neighbors=1)
model2 = KNeighborsClassifier(n_neighbors=3)


In [0]:
from sklearn.model_selection import cross_val_score
scores1 = cross_val_score(model1, X, y, cv=5, scoring = 'accuracy')
scores2 = cross_val_score(model2, X, y, cv=5, scoring = 'accuracy')
print('scores1 : ', scores1.mean())
print('scores2 : ', scores2.mean())


In [160]:
linsvc = LinearSVC()
linsvc.fit(one_train[final_col],one_train['isFraud'])
linsvc_score = round(linsvc.score(one_train[final_col],one_train['isFraud'])*100, 2)

linsvc_score

/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


100.0

In [161]:
print(linsvc_score)

100.0


In [0]:
def reduce_mem_usage(df):
    start_mem = df.memory_usage().sum() / 1024**2
    for col in df.columns:
        col_type = df[col].dtype
        
        if col_type != object:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[: 3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min  and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
    
    end_mem = df.memory_usage().sum()/ 1024**2
    print('Memory usage after optimization is: {:.2f} MB, {:.1f}% reduction'.\
          format(end_mem, 100 * (start_mem - end_mem) / start_mem))
    
    return df

In [157]:
one_train = reduce_mem_usage(one_train)

Memory usage after optimization is: 248.09 MB, 70.5% reduction


In [0]:
print(round(train_transaction['isFraud'].value_counts(normalize=True) * 100,2))
train_transaction['isFraud'].astype(int).plot.hist();